File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.cpp
class Solution {
public:
    int strStr(string ha, string ne) {
        if(ne.size()>ha.size()){
            return -1;
        }
        else if(ne.size()==ha.size()){
            if(ne!=ha){
                return -1;
            }
            return 0;
        }
        int res=-1;
        string temp="";
        for(int i=0;i<ne.size();++i){
            temp+=ha[i];
        }
        if(temp==ne){
            return 0;
        }
        int s=1;
        for(int i=ne.size();i<ha.size();++i){
            temp="";
            for(int j=s;j<=i;++j){
                temp+=ha[j];
            }
            // cout<<temp<<endl;
            if(temp==ne){
                return s;
            }
            ++s;
        }
        return -1;
    }
};

File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.py
class Solution(object):
    def strStr(self, ha, ne):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        if len(ne)>len(ha):
            return -1
        elif len(ne)==len(ha):
            if ne!=ha:
                return -1
            return 0
        p=7
        hss=0
        hs=0
        for i in range(len(ne)):
            hss+=(ord(ne[i])-96)*int(p**(i+1))
            hs+=(ord(ha[i])-96)*int(p**(i+1))
        if hss==hs:
            return 0
        for i in range(len(ne),len(ha),1):
            hs-=(ord(ha[i-len(ne)])-96)*(p**(i-len(ne)+1))
            hs+=(ord(ha[i])-96)*(p**(i+1))
            if hss==hs//(p**(i-len(ne)+1)):
                return i-len(ne)+1
        return -1
        